File: leetcode/Algorithms/131.PalindromePartitioning/Solution.py
class Solution(object):
    def partition(self, s):
        """
        :type s: str
        :rtype: List[List[str]]
        """
        cache = [[]]

        for index in range(len(s) - 1, -1, -1):
            tmp, suffix = [], s[index:]

            for cursor in range(index + 1, len(s)):
                sub = s[index:cursor]
                if sub == sub[::-1]:
                    for this_list in cache[len(s) - cursor]:
                        next_list = this_list[:]
                        next_list.insert(0, sub)
                        tmp.append(next_list)

            if suffix == suffix[::-1]:
                tmp.append([suffix])

            cache.append(tmp)

        return cache[-1]

File: leetcode/Algorithms/131.PalindromePartitioning/Solution.java
import java.util.ArrayList;
import java.util.List;

public class Solution {
    private List<List<String>> result = new ArrayList<>();
    private List<String> tmp = new ArrayList<>();

    public List<List<String>> partition(String s) {
        backtracking(s, 0);
        return result;
    }

    private void backtracking(String s, int index) {
        if (index == s.length()) {
            result.add(new ArrayList<>(tmp));
        } else {
            for (int i = index; i < s.length(); i++) {
                String prefix = s.substring(index, i + 1);
                if (isPalindrome(prefix)) {
                    tmp.add(prefix);
                    backtracking(s, i + 1);
                    tmp.remove(tmp.size() - 1);
                }
            }
        }
    }

    private boolean isPalindrome(String string) {
        int start = 0, end = string.length() - 1;
        while (start < end) {
            if (string.charAt(start++) != string.charAt(end--)) return false;
        }
        return true;
    }
}
